src/ethereum/frontier/vm/gas.py

```diff
@@ -45,3 +48,56 @@ def subtract_gas(gas_left: U256, amount: U256) -> U256:
         raise OutOfGasError
 
     return gas_left - amount
+
+
+def calculate_memory_gas_cost(size_in_bytes: U256) -> U256:
+    """
+    Calculates the gas cost for allocating memory up
+    and including the given address.
+
+    Parameters
+    ----------
+    size_in_bytes :
+        The size of the data in bytes.
+
+    Returns
+    -------
+    total_gas_cost : `ethereum.base_types.U256`
+        The gas cost for storing data in memory.
+    """
+    size_in_words = ceil32(size_in_bytes) // 32
+    linear_cost = size_in_words * GAS_MEMORY
+    quadratic_cost = size_in_words ** 2 // GAS_MEMORY_QUADRATIC_DENOMINATOR
+    total_gas_cost = linear_cost + quadratic_cost
+    return total_gas_cost
+
+
+def calculate_gas_for_extending_memory(
```

Comment:
What does this return if you try to `mstore` a bit of memory which already exists? Something like:

```python
from ethereum.frontier.vm.gas import calculate_gas_for_extending_memory 

memory = bytearray(b"\x00" * 256)
calculate_gas_for_extending_memory(memory, 0, 32)
```

I suspect it tries to return a negative number?

Reply:
@SamWilsn Nice catch Sam! I'll add a fix for this.

Reply:
fixed.